### PR TITLE
[SLS-1671] Inferred Spans for Kineses, EventBridge, S3, DynamoDB

### PR DESCRIPTION
**Pull request by hghotra**
### What does this PR do?

Add inferred Spans for Kineses, EventBridge, S3, & DynamoDB

**Kinesis**

image

**EventBridge**

image

**S3**

image

**DynamoDB**

image

### Motivation

https://datadoghq.atlassian.net/browse/SLS-1671

**Context:** https://docs.google.com/document/d/1g50vbbDA-yog5RdQsRE8YAQ546Ve7J9inzn64S-ars8/edit#heading=h.mog4x98ex87z

### Testing Guidelines

Run unit tests:

`./scripts/run_tests.sh`

Run integration tests:

`BUILD_LAYERS=true DD_API_KEY=<api key here> aws-vault exec sandbox-account-admin -- ./scripts/run_integration_tests.sh`

Run the inferred spans example in the cloud-integrations-lambdas repo.

DataDog/cloud-integrations-lambdas@6922179

### Types of Changes

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [x] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [x] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [ ] This PR passes the integration tests (ask a Datadog member to run the tests)

### CODE REVIEW

File: datadog_lambda/tracing.py [hunk 1]
```diff
@@ -412,13 +412,26 @@ def create_inferred_span(event, context):
...
+            logger.debug("Eventbridge event detected. Inferring a span")
+            return create_inferred_span_from_eventbridge_event(event, context)
 
     except Exception as e:
         logger.debug(
             "Unable to infer span. Detected type: {}. Reason: {}",
             event_source.to_string(),
             e,
         )
+        print(e)
```
Comment (astuyve) [inline datadog_lambda/tracing.py:434]: do we need this and the logger.debug?

Comment (hghotra) [reply datadog_lambda/tracing.py]: Sorry, that sneaked in there with the evenBridge change, ✂️

File: datadog_lambda/tracing.py [hunk 2]
```diff
@@ -412,13 +412,26 @@ def create_inferred_span(event, context):
         elif event_source.equals(EventTypes.SNS):
             logger.debug("SNS event detected. Inferring a span")
             return create_inferred_span_from_sns_event(event, context)
+        elif event_source.equals(EventTypes.KINESIS):
```
Comment (astuyve) [inline datadog_lambda/tracing.py:415]: long term we might consider a null object pattern for event detection. Not necessary for this PR, just a suggestion

**Review by astuyve — APPROVED** [review]

File: datadog_lambda/tracing.py (hunk 2 above)
Comment (agocs) [reply datadog_lambda/tracing.py:415]: Agreed. One of the reasons this was implemented like this is that the original trigger detection code handles some known events, but also allows "known-unknown" events that just have a raw string in the "source" or "event_source" key. I think the direction in which we're moving is a little bit of a departure from Postel's law. Thats not a bad thing, but we want to make sure we don't lose any functionality.

**Pull request merged**